Refactor Language (#4)
nidhijaju/draft-zstd-window-size: draft-jaju-httpbis-zstd-window-size.md

--- abstract

Deployments of Zstandard, or "zstd", can use different window sizes to limit
memory usage during compression and decompression. Some browsers and user
agents limit window sizes to mitigate memory usage concerns, causing
interoperability issues. This document updates the window size limit in RFC8878
from a recommendation to a requirement in HTTP contexts.

--- middle

# Introduction

Zstandard, or "zstd", specified in {{?RFC8878}}, is a lossless data compression
mechanism similar to gzip. When used with HTTP, the "zstd" Content Coding
token signals to the decoder that the content is Zstandard-compressed.

> An important property of Zstandard-compressed content is its Window_Size
> ({{!RFC8878, Section 3.1.1.1.2}}), which describes the maximum distance for
> matches and therefore how much of the content must be kept in memory during
> decompression. Larger Window_Size values tend to improve the compression ratio,

Reviewer: ```suggestion
Deployments of Zstandard use different Window_Size values
({{!RFC8878, Section 3.1.1.1.2}}) to describe the maximum back-reference distance,
which controls the amount of content that must be kept in memory during
decompression. Larger Window_Size values tend to improve the compression ratio,
```
Proposed a diff that combines the two approaches with somewhat cleaner phrasing.
People can probably guess at the meaning of "matches", but maybe "back-references" is better since it's also used in RFC8878?

Author: The existing/proposed phrase

> Deployments of Zstandard use different Window_Size values [...]

is slightly inaccurate: the `Window_Size` is a property of a frame. Deployments/integrations of Zstandard may or may not constrain the `Window_Size`s they accept or produce, but that's a slightly different thing.

Maybe that's too pedantic a distinction and we can be a little looser with the language. But in my PR here I've tried to handle that nuance by first describing the `Window_Size` property of frames in this paragraph and then in the next talking about how deployments interact with it. I can make that second paragraph clearer perhaps?

> during decompression. The minimum Window_Size is 1 KB. The maximum Window_Size
> is (1<<41) + 7*(1<<38) bytes, which is 3.75 TB. Larger Window_Size values tend
> to improve the compression ratio, but at the cost of increased memory usage.

Reviewer: ```suggestion
during decompression.

The minimum Window_Size is 1 KB. The maximum Window_Size is
(1<<41) + 7*(1<<38) bytes, which is 3.75 TB. Larger Window_Size values tend
to improve the compression ratio, but at the cost of increased memory usage.
```
Editorial nit: this just improves readability

To protect against unreasonable memory usage, some browsers and user agents

> limit the maximum Window_Size they will handle. Since the "zstd" Content Coding
> does not specify a negotiation mechanism for the Window_Size to use, this

Reviewer: ```suggestion
limit the maximum Window_Size they will handle. This
```
This has caused some controversy in the WG in the past, so let's keep negotiation mechanisms out of scope here.

Author: Sounds good.

causes failures to decode responses if the content is compressed with a larger

> Window_Size than the user agent allows.

Reviewer: ```suggestion
Window_Size than the user agent allows, leading to decreased interoperability.
```
Ensuring interoperability is the reason we wrote this draft so let's keep that phrasing here.

Author: Sounds good.

{{!RFC8878, Section 3.1.1.1.2}} recommends that decoders support a Window_Size
of up to 8 MB, and that encoders not generate frames using a Window_Size larger
than 8 MB. However, it imposes no requirements.

> This document amends the specification of the "zstd" Content Coding in

Reviewer: We should keep the above sentence with "updates {{RFC8878}}" since updating an RFC is an official process, whereas "amends the specification" doesn't have the same meaning.

Author: Aha, ok. Will do.

# Window Size

Section 3.1.1.1.2 of {{RFC8878}} discusses window sizes in Zstandard.
The window size provides guarantees about the minimum memory buffer required to
decompress a frame. This information is important for decoders to allocate
enough memory.

The minimum window size is 1 KB. The maximum window size is (1<<41) +
7*(1<<38) bytes, which is 3.75 TB.

In general, larger window size values tend to improve the compression ratio, but
at the cost of increased memory usage.

To properly decode compressed data, a decoder will need to allocate a buffer of
at least the window size bytes.

In order to protect decoders from unreasonable memory requirements, a decoder is
allowed to reject a compressed frame that requests a memory size beyond the

> decoder's authorized range.

Reviewer: I think it helps to keep some variation of this text here for reading comprehension, especially since the intention of this document is to be the reference point for window size for zstd.
